equinox/nn/attention.py

```diff
@@ -11,6 +11,183 @@
 from .linear import Linear
 
 
+def dot_product_attention_weights(
+    query: Array,
+    key_: Array,
+    bias: Optional[Array] = None,
+    mask: Optional[Array] = None,
+    dropout_p: float = 0.0,
+    *,
+    dropout: Optional[Dropout] = None,
+    key: Optional["jax.random.PRNGKey"] = None,
+    inference: Optional[bool] = None,
+    deterministic: Optional[bool] = None,
+):
+
+    r"""Computes dot-product attention weights given query and key.
+    Used by :func:`dot_product_attention`, which is what you'll most likely use.
+    But if you want access to the attention weights for introspection, then
+    you can directly call this function and call einsum yourself.
+    This also supports multi-query attention (https://arxiv.org/pdf/1911.02150).
+    **Arguments:**
+
+        - `query`: Query vectors. Should be a JAX array of shape
+            `(batch ... query_seq_length, num_heads, qk_size)`.
+        - `key_`: Key vectors. Should be a JAX array of shape
+            `(batch ... kv_seq_length, num_heads, qk_size)`.
+        - `mask`: Optional mask preventing attention to certain positions. Should be a
+            JAX array of shape `(num_heads, query_seq_length, kv_seq_length)`.
+        - `dropout_p`: Dropout probability on attention weights.
+        - `dropout`: Already initialized Dropout module. Unused if `dropout_p != 0.`
+          (Keyword only argument).
+        - `key`: A `jax.random.PRNGKey` used for dropout. Unused if `dropout_p = 0.`.
+            (Keyword only argument.)
+        - `inference`: As [`equinox.nn.Dropout.__call__`][]. (Keyword only
+            argument.)
+        - `deterministic`: (Deprecated in favour of `inference`.)
+
+        **Returns:**
+
+        A JAX array of shape `(batch..., num_heads, query_seq_length, kv_seq_length)`.
+    """
+    if query.ndim == key_.ndim:
+        assert query.shape[:-3] == key_.shape[:-3], "q, k batch dims must match."
+        assert query.shape[-2] == key_.shape[-2], "q, k num_heads must match."
+        assert query.shape[-1] == key_.shape[-1], "q, k depths must match."
+    elif query.ndim > key_.ndim:
+        # support for multi-query attention
+        assert query.shape[:-3] == key_.shape[:-2], "q, k batch dims must match."
+        assert query.shape[-1] == key_.shape[-1], "q, k depths must match."
+    else:
+        raise ValueError("q must have equal or more dimensions than k.")
+
+    query_seq_length, num_heads, depth = (
+        query.shape[-3],
+        query.shape[-2],
+        query.shape[-1],
+    )
+    kv_seq_length = key_.shape[-2] if key_.ndim < query.ndim else key_.shape[-3]
+
+    query = query / jnp.sqrt(depth)
+    if query.ndim == key_.ndim:
+        attn_weights = jnp.einsum("...qhd,...khd->...hqk", query, key_)
+    else:
+        # multi-query attention
+        attn_weights = jnp.einsum("...qhd,...kd->...hqk", query, key_)
```

Comment:
I think multi-query attention is something you've added in here, i.e. it wasn't supported by the original implementation? I'm a bit skeptical about merging two different operations together like this. I think this might be possible to support more neatly by writing this function as operating on single-headed attention, and then either `vmap`'ing over both query and key (=normal multihead attention) or just `vmap`'ing over query (=multi-query attention)?

Reply:
Yeah it actually probably doesn't make sense to add support for multiquery attention in this PR - I can make a follow-up that adds it in so that this one only handles straightforward multihead attention utilities.

Reply:
Went ahead and removed this for now.